controllers/utils/failuredomains.go

```diff
@@ -0,0 +1,139 @@
+			if !strings.HasSuffix(fdSpec.Name, "-"+r.CAPICluster.Name) { // Add cluster name suffix if missing.
+				fdSpec.Name = fdSpec.Name + "-" + r.CAPICluster.Name
+			}
+			if err := r.CreateFailureDomain(fdSpec); err != nil {
+				if !strings.Contains(strings.ToLower(err.Error()), "already exists") {
+					return reconcile.Result{}, errors.Wrap(err, "creating CloudStackFailureDomains")
+				}
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+// GetFailureDomains gets CloudStackFailureDomains owned by a CloudStackCluster.
+func (r *ReconciliationRunner) GetFailureDomains(fds *infrav1.CloudStackFailureDomainList) CloudStackReconcilerMethod {
+	return func() (ctrl.Result, error) {
+		capiClusterLabel := map[string]string{clusterv1.ClusterLabelName: r.CSCluster.GetLabels()[clusterv1.ClusterLabelName]}
+		if err := r.K8sClient.List(
+			r.RequestCtx,
+			fds,
+			client.InNamespace(r.Request.Namespace),
+			client.MatchingLabels(capiClusterLabel),
+		); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to list failure domains")
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+// GetFailureDomainsAndRequeueIfMissing gets CloudStackFailureDomains owned by a CloudStackCluster and requeues if none are found.
+func (r *ReconciliationRunner) GetFailureDomainsAndRequeueIfMissing(fds *infrav1.CloudStackFailureDomainList) CloudStackReconcilerMethod {
+	return func() (ctrl.Result, error) {
+		if res, err := r.GetFailureDomains(fds)(); r.ShouldReturn(res, err) {
+			return res, err
+		} else if len(fds.Items) < 1 {
+			return r.RequeueWithMessage("no failure domains found, requeueing")
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+type CloudClientExtension interface {
+	RegisterExtension(*ReconciliationRunner) CloudClientExtension
+	AsFailureDomainUser(*infrav1.CloudStackFailureDomainSpec) CloudStackReconcilerMethod
+}
+
+type CloudClientImplementation struct {
+	CloudClientExtension
+	*ReconciliationRunner
+}
+
+func (c *CloudClientImplementation) RegisterExtension(r *ReconciliationRunner) CloudClientExtension {
+	c.ReconciliationRunner = r
+	return c
+}
+
+// AsFailureDomainUser uses the credentials specified in the failure domain to set the ReconciliationSubject's CSUser client.
+func (c *CloudClientImplementation) AsFailureDomainUser(fdSpec *infrav1.CloudStackFailureDomainSpec) CloudStackReconcilerMethod {
+	return func() (ctrl.Result, error) {
+		if !strings.HasSuffix(fdSpec.Name, "-"+c.CAPICluster.Name) { // Add cluster name suffix if missing.
+			fdSpec.Name = fdSpec.Name + "-" + c.CAPICluster.Name
+		}
+		endpointCredentials := &corev1.Secret{}
+		key := client.ObjectKey{Name: fdSpec.ACSEndpoint.Name, Namespace: fdSpec.ACSEndpoint.Namespace}
+		if err := c.K8sClient.Get(c.RequestCtx, key, endpointCredentials); err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "getting ACSEndpoint secret with ref: %v", fdSpec.ACSEndpoint)
+		}
+
+		var err error
+		if c.CSClient, err = cloud.NewClientFromK8sSecret(endpointCredentials); err != nil {
```

Comment:
should we use pointers for clients and consider some caching mechanisms?

Reply:
Yeah, I don't hate the idea.

Still even if we cache the clients, I'm not sure how that would act. The clients themselves will run in a new thread for every reconciliation request. I'm wondering if the clients would try to share a connection there or not.

Either way, I do want to look into efficiency of connections here, so I'll do some research.